src/task.rs

```diff
@@ -396,37 +351,76 @@ impl Task {
             return self.notify_starling_errored(Error::from_kind(ErrorKind::JavaScriptException));
         }
 
-        // Drain the micro-task queue.
-
         unsafe {
             jsapi::js::RunJobs(cx);
+        }
 
-            if jsapi::JS_IsExceptionPending(cx) {
+        if let Err(e) = self.check_for_unhandled_rejected_promises() {
+            return self.notify_starling_errored(e);
+        }
+
+        self.evaluate_main()
+    }
+
+    fn evaluate_main(&mut self) -> Result<Async<()>> {
+        let cx = self.runtime().cx();
+        rooted!(in(cx) let global = self.global.get());
+
+        let mut has_main = false;
+        unsafe {
+            assert!(jsapi::JS_HasProperty(
+                cx,
+                global.handle(),
+                b"main\0".as_ptr() as _,
+                &mut has_main
+            ));
+        }
+        if !has_main {
+            return self.notify_starling_finished();
+        }
+
+        rooted!(in(cx) let mut rval = jsval::UndefinedValue());
+        let args = jsapi::JS::HandleValueArray::new();
+        unsafe {
+            let ok = jsapi::JS_CallFunctionName(
+                cx,
+                global.handle(),
+                b"main\0".as_ptr() as _,
+                &args,
+                rval.handle_mut()
+            );
+
+            if !ok {
+                debug_assert!(jsapi::JS_IsExceptionPending(cx));
                 // TODO: convert the pending exception into a meaningful error.
                 jsapi::JS_ClearPendingException(cx);
+                jsapi::js::RunJobs(cx);
                 return self.notify_starling_errored(
                     Error::from_kind(ErrorKind::JavaScriptException)
                 );
             }
-        }
 
-        // Check for unhandled, rejected promises.
-
-        let unhandled = {
-            let mut tracker = self.rejected_promises.borrow_mut();
-            tracker.take()
-        };
+            jsapi::js::RunJobs(cx);
+        }
 
-        if !unhandled.is_empty() {
-            // TODO: collect the rejection values from the promise, and
-            // line/column/etc if the rejection value is some kind of Error
-            // object.
-            return self.notify_starling_errored(Error::from_kind(
-                ErrorKind::JavaScriptUnhandledRejectedPromise,
-            ));
+        if let Err(e) = self.check_for_unhandled_rejected_promises() {
+            return self.notify_starling_errored(e);
         }
 
-        self.notify_starling_finished()
+        rooted!(in(cx) let mut obj = ptr::null_mut());
+        if unsafe {
```

Comment:
Woah, this threw me for a loop. Is this kinds of code common and accepted practice? If so, learn something new everyday, I guess. If not, please change it to assign the result of the unsafe block to a local and then use that in the `if` condition.

Reply:
Certainly is a thing in other expression-oriented languages, but Rust is *both* one of those and an imperative, mutation-oriented language. I can change it.